File: app/services/census.service.ts
import { firebase } from '@nativescript/firebase-core';
import '@nativescript/firebase-database';
import { CensusEntry } from '../models/census.model';

export class CensusService {
    private static instance: CensusService;
    
    static getInstance(): CensusService {
        if (!CensusService.instance) {
            CensusService.instance = new CensusService();
        }
        return CensusService.instance;
    }

    async addEntry(entry: Omit<CensusEntry, 'id'>): Promise<string> {
        const database = firebase().database();
        const newEntryRef = database.ref('census').push();
        
        await newEntryRef.set({
            ...entry,
            id: newEntryRef.key
        });
        
        return newEntryRef.key;
    }

    async getEntriesByVotingCenter(votingCenter: string): Promise<CensusEntry[]> {
        const database = firebase().database();
        const snapshot = await database
            .ref('census')
            .orderByChild('votingCenter')
            .equalTo(votingCenter)
            .once('value');
            
        const entries: CensusEntry[] = [];
        snapshot.forEach((childSnapshot) => {
            entries.push(childSnapshot.val() as CensusEntry);
        });
        
        return entries;
    }
}

File: app/src/main/kotlin/org/censusapp/services/CensusService.kt
import com.google.firebase.database.FirebaseDatabase
import kotlinx.coroutines.tasks.await

class CensusService private constructor() {
    private val database = FirebaseDatabase.getInstance()

    suspend fun addEntry(entry: CensusEntry): String {
        val ref = database.reference.child("census").push()
        val id = ref.key ?: throw IllegalStateException("Failed to generate key")
        
        ref.setValue(entry.copy(id = id)).await()
        return id
    }

    suspend fun getEntriesByVotingCenter(votingCenter: String): List<CensusEntry> {
        val snapshot = database.reference
            .child("census")
            .orderByChild("votingCenter")
            .equalTo(votingCenter)
            .get()
            .await()

        return snapshot.children.mapNotNull { it.getValue(CensusEntry::class.java) }
    }

    companion object {
        @Volatile
        private var instance: CensusService? = null

        fun getInstance() = instance ?: synchronized(this) {
            instance ?: CensusService().also { instance = it }
        }
    }
}